pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+
+        the module's content is accessible via node.stream() function
+        """
+        codec: Optional[str]
+        lines: Sequence[bytes]
+        invalid_codec_line: Optional[int] = None
+
+        # First try to detect encoding with PEP 263, won't work
+        # see https://bugs.python.org/issue1503789
+        # TO DO we could manually modify the stream to convert UTF16/UTF32 to ASCII
+        #      but why ;-) ?
+        try:
+            with node.stream() as stream:
+                codec, lines = detect_encoding(stream.readline)
+        except SyntaxError:
+            pass
+        else:
+            if self.is_invalid_codec(codec):
+                invalid_codec_line = len(lines)
+
+        # If there is no invalid codec already, lets try to determine a BOM,
+        # even so Python currently doesn't like non UTF-8 BOMs ....
+        if invalid_codec_line is None:
+            with node.stream() as stream:
+                codec = extract_codec_from_bom(stream.readline())
+            if self.is_invalid_codec(codec):
+                invalid_codec_line = 1
+
+        if invalid_codec_line is not None:
+            self.add_message(
+                "invalid-unicode-codec",
+                line=invalid_codec_line,
+                confidence=interfaces.HIGH,
+                col_offset=0,
+            )
+
+
+class BidirectionalUnicodeChecker(BaseChecker):
+    """
+    Check Unicode that could be used to hide bad code to humans
+    """
+
+    # Based on:
+    # https://golangexample.com/go-linter-which-checks-for-dangerous-unicode-character-sequences/
+    BAD_UNICODE = [
+        "\u202A",
+        "\u202B",
+        "\u202C",
+        "\u202D",
+        "\u202E",
+        "\u2066",
+        "\u2067",
+        "\u2068",
+        "\u2069",
+        # The following was part of the PEP 672 project:
+        # https://www.python.org/dev/peps/pep-0672/
+        # so the list above might not be complete
+        "\u200f",
+        # We don't use "\u200e" (LEFT to RIGHT MARK)
+        # as this is the default for latin files and can't be used
+        # to hide code
+    ]
+
+    __implements__ = interfaces.IRawChecker
+    priority = -1
+
+    name = "bidirectional_unicode_checker"
+    msgs = {
+        "E1801": (
+            "bidirectional unicode can be used to hide real code execution",
+            "bidirectional-unicode",
+            "bidirectional unicode can be used to hide real code execution",
+        )
+    }
+    options = ()
+
+    def process_module(self, node: nodes.Module) -> None:
+        """
+        Perform the the actual checking by checking module stream
+
+        the module's content is accessible via node.stream() function
+        """
+        with node.stream() as stream:
+            for (lineno, line) in enumerate(stream):
+                line_contains_error: bool = False
+                for dangerous in self.BAD_UNICODE:
+                    # We assume that the file is encoded in UTF-8!
```

Comment:
I'm not sure we can assume this. What would happen if somebody disabled the `invalid-unicode-codec` checker?

Reply:
Currently Nothing, as it is not possible in Python at the moment to use the UTF16/UTF32 at all. And given to some expression in the python bug report that I linked it is also not gonna happen in the near future.

In other non UTF codecs there is no problem with bidi codecs.

Problem is that it is damn near impossible to write some Checker for this cases, as the AST Parser will simply crash at the moment.
So IMO we are fine with the invalid-unicode-codec checks. Allowing to check for bidirectional unicode also in UTF-16/UTF-32 would increase the complexity by a maginitude!
The Problem is that UTF-16 und UTF-32 are fixed with, so the byte representation of the bidi codes represent valid ASCII. 
And we are not able to decode the byte stream, as we don't know how Python would implement the codec check.
So kinda of a chicken egg problem.

Reply:
Understandable and okay?
If yes please resolve.

Reply:
Outdated: Due to combining the Checkers, I don't have to assume UTF-8 anymore :-) @DanielNoord